### DISCUSSION TITLE
Feature request: Make UI scaling and view scaling independent of one another

### DISCUSSION THREAD

**Discussion by DanielWeiner** [endless-sky/endless-sky#8071] [Ideas]
I have a 1440p monitor and I like to have the UI scaled to 200%. It makes the text a lot more readable to me. The only problem with that is that it affects how far I can zoom out while in space. With main zoom at 200% I get a very small view of my surroundings when zoomed out all the way, and I'd like to have the same range of vision that I'd get if main zoom was at 100%. Being too zoomed in makes combat a bit more difficult to manage.

**Comment by Zitchas** [comment]
I agree. Needing things to be bigger in order to read or make out UI details shouldn't be affecting the main space screen.

**Comment by vitalchip** [comment]
Do they  really need to be independent? Sounds like you just need an extra level of zoom out, so you can see a larger area of space?

**Reply by Zitchas** [reply]
Sure, having the main "action" area zoom have additional levels would be good.

**Reply by DanielWeiner** [reply]
An additional zoom level would also be good, but it wouldn't address the root of the issue. You'd still end up with different levels of zoom depending on your UI scale.

**Reply by vitalchip** [reply]
Sure, I know they don't all have exact opposites.  100/100 on my screen gives 840px centre - edge, for example.  If I set main zoom to 110/120/130 there's no view zoom level which keeps the same screen width. The closest is 70 but it gives 900 - 1000px width for those main zooms.  You need to go 140/70 to get back to the 840px (give or take a couple of px)

So, they could be better matched but how much does it really matter?  When would it be essential that the screen remained exactly 840px and it being 900px at a different combination was no good?

**Reply by DanielWeiner** [reply]
End of the day, it probably doesn't matter _too_ much. I'm just personally, as a developer, a stickler against unnecessary coupling. But as a player what I care about is: 

* Having the UI scaled to 200% so that it's easy to read
* Being able zoom out enough to be able to see what's going on during combat so I can judge my opportunity to strike
* Being able to go to the map screen and see the names of the systems without zooming in too far.

How that's implemented, it's just a matter of preference. End of the day, I'm not (yet) a contributor so take my ideas on implementation with a grain of salt. But those are the things I'd like to have as a player.

**Reply by Zitchas** [reply]
Well, we always need more coders, if you have the interest... :)

**Reply by vitalchip** [reply]
The reason I ask is that, generally speaking I find them being coupled to be more beneficial than if they were not. When I use main zoom I want everything to be larger at the top end of the view zoom range. As long as I can zoom out far enough.

Each option is more or less the same though, it's either they move together and sometimes you need to zoom out or they are independent and sometimes we'll need more zoom in.

Adding extra levels of zoom is trivial, making the two zooms truly independent (or adding a seperate zoom for Ui elements) is likely to be much more involved.

**Comment by vitalchip** [comment]
The vector of zoom levels is in Preferences.cpp.

const vector<double> ZOOMS = {.25, .35, .50, .70, 1.00, 1.40, 2.00}; 

Just chucking in an extra couple of levels at the lower end would probably cover your initial needs.

const vector<double> ZOOMS = {.125, .175, .25, .35, .50, .70, 1.00, 1.40, 2.00, 2.80};

I'm currently using:

const vector<double> ZOOMS = {.05, .1, .2, .3, .4, .5, .6, .7, .80, .90, 1.00, 1.10, 1.20, 1.30, 1.40, 1.50, 1.60, 1.70, 1.80, 1.90, 2.00, 3.00, 4.00};  for no particular reaason.

The highest zoom I mostly use when I need to see hardpoints up close and the lower two levels switch to a sort of tactical overview where I can see about 40,000px worth of system activity.


![Screenshot 2023-01-06 at 20 20 53](https://user-images.githubusercontent.com/65342577/211113387-876a4f3a-cda5-4b68-8e3b-99765ed6bcbf.png)

**Reply by Zitchas** [reply]
Is that showing your landing target on the left; with the distance to it?

And those are status effect indicators on the right?

You've got so much stuff on that screen that really intrigues me.

**Reply by Zitchas** [reply]
What does the "zoomIndex = 4" do?

**Reply by warp-core** [reply]
> What does the "zoomIndex = 4" do?

If it's where I think it is, it sets the zoom setting to the item at index 4 (the fifth element) of whatever collection it's looking at (presumably, the collection corresponding to zoom values).

**Reply by Zitchas** [reply]
Is your build on GitHub? Given all the changes you've got and talked about, I'd love to take that for a spin sometime.

**Reply by Zitchas** [reply]
I don't have time to test it right now. Hopefully will later, but it looks like it's building now:

https://github.com/Zitchas/endless-sky/pull/132

**Reply by vitalchip** [reply]
I only have a local build now and it's old code too.  9.12 and some 9.13.   It's well out of date but has a lot of 'experiments' added. There hasn't been much added to the code since that I didn't already have or just didn't want.  I play about with it from time to time but it's just tinkering when I'm at a loose end these days.,

**Reply by Zitchas** [reply]
Ah. Well, it'd still be fun to try sometime.

**Reply by DanielWeiner** [reply]
Niiiiiiiiiice, you folks are awesome

**Reply by Zitchas** [reply]
I took the time to test it anyway.
a) it works, and
b) looks good!

The background haze and asteroids tile a bit, but other than that, very nice. Neat being able to watch the whole system like that!

(I implemented the full set that Vitalchip posted, and I think I even managed to get the whole split onto two lines thing to work...)

**Reply by vitalchip** [reply]
Yeh, the haze is pretty messed up when zoomed out all the way and it runs the GPU a lot too (for me anyway) I just don't display them at those levels. I would imagine the tiled asteroids look bad too but I got rid of them a long time ago.

**Comment by vitalchip** [comment]
Is that showing your landing target on the left; with the distance to it?

And those are status effect indicators on the right?

You've got so much stuff on that screen that really intrigues me.

Yes, that's the planet I'm targeting for landing, easier to see than the messages along the bottom, particularly if you switch landing targets. On the right is a speedo, slowing percentage, solar collection and incoming missile count.

The three horizontal bars are what's being commanded of the various thrusters.

**Reply by Zitchas** [reply]
That's so much useful into... I'd love to have all that in our GUI. *Especially* for development and testing.

**Reply by vitalchip** [reply]
A lot of it was added at various points when I was trying things out and generally wondering what was going wrong!  I have a similar lot of stuff available for targeted ships.

<img width="285" alt="Screenshot 2023-01-06 at 23 47 19" src="https://user-images.githubusercontent.com/65342577/211118342-367c5f3f-5877-4da1-b68d-fd777edadc0b.png">
<img width="286" alt="Screenshot 2023-01-06 at 23 46 44" src="https://user-images.githubusercontent.com/65342577/211118354-d27fb31f-2c5f-4326-84cd-653cb4fb0a10.png">

**Reply by Zitchas** [reply]
I don't know if I could get that into vanilla; but that would be *awesome* for debug; and honestly I'd probably just play my own version of the game just to keep that sort of stuff. Is it attached to the tactical scan? That'd really make the tactical scan useful, finally.

**Reply by vitalchip** [reply]
Yes, the gun and turret ranges and the turn speed (after slowing) are part of the tactical display. The text below is pure testing text and stays visible.

**Comment by DanielWeiner** [comment]
Here's an update: I decided to take a dive and actually try implementing this myself. My proof-of-concept is pretty rough locally, but it does what I want it to to. Might make a PR once I'm fully satisfied with the change

**Reply by DanielWeiner** [reply]
![image](https://user-images.githubusercontent.com/3599467/212430395-b9b2cc54-5559-406f-80e0-55027a2fefd0.png)
